examples/systemcalls/systemcalls.c

```diff
@@ -92,6 +146,51 @@ bool do_exec_redirect(const char *outputfile, int count, ...)
  *   The rest of the behaviour is same as do_exec()
  *
 */
+    pid_t c_pid;
+    int fd;
+    int w_status;
+
+    fd = open (outputfile, O_WRONLY|O_TRUNC|O_CREAT, 0644);
+
+    if (fd < 0)
+	    return -1;
+
+    c_pid = fork();
+    switch (c_pid)
+    {
+    	case -1: 
+		return -1;
+	case 0:
+		if (dup2(fd,1) <0)
+		{
+			close(fd);
+			exit(EXIT_FAILURE);
+		}
+		
+		int ret = execv(command[0], command);
+        	if (ret == -1)
+			exit (-1);
+	default:
+		 close(fd);
+	         if (waitpid (c_pid, &w_status, 0) == -1)
+		 {
+		        return -1;
+		 }
+    		
+		 if (WIFEXITED (w_status))
+		 {
+		 	if (WEXITSTATUS (w_status) != 0)
+			{
+				return -1;
+			}
+
+		 }
+
+    		else
+		{
+			return -1;
```

Comment:
Missing cleanup in these code paths.
Also see discussion in class about avoiding duplication between these two very similar functions.